Disabled bsync function by default, added function to enable it.

andrewbenton opened the pull request:
By default, barrier sync functionality is disabled by default with the PR.  By calling setCanBarrier(), the barrier sync can be re-enabled on a test-wise level.  In the case that the barrier sync is not enabled for a test, and one of that test's glue functions calls Barrier.sync(), Zucchini will throw a ZucchiniRuntimeException.

andrewbenton committed:
Force barrier to be enabled for use.

Previously, all tests had the framework for the barrier sync
automatically enabled.  This change makes it so that the barrier sync
functionality needs to be enabled in order to be used, otherwise it will
skip barrier sync functions and throw if the barrier sync is called when
not supported.  The barrier sync is now disabled by default.

andrewbenton committed:
Merge branch 'develop' of github.com:andrewbenton/zucchini into develop

andrewbenton committed:
Skipped area that could lead to undefined behavior

andrewbenton committed:
Saving bad syntax after test not suggested practice.

dimaj commented on src/main/java/com/comcast/zucchini/AbstractZucchiniTest.java:
```diff
@@ -132,6 +147,8 @@ private boolean envSerialized() {
      * @param contexts The list of contexts that will be run.
      */
     public void runParallel(List<TestContext> contexts) {
+        this.canBarrierInTest = this.canBarrier;
```
instead of setting this property from both runParallel and runSerial, why not call it once from run?

andrewbenton replied:
With the API, runParallel() and runSerial() are both public methods.  Since they are public, the run() method is not always called.  In the case that someone decided to call the bare runParallel() and runSerial() methods, this method prevents bad behavior on the part of zucchini.  We _could_ put something in the docs about undefined behavior on using a barrier and calling the bare runParallel() and runSerial() methods, and then we could put the statement just in the run() method, but I think that's less stable.  

tl;dr - The current method of doing this is the belt-and-suspenders approach.  For the next api change, it might desirable to make the runParallel() and runSerial() private, and provide an overloaded run() method so that it defaults in accordance with envSerialized(), and can be overridden with run(boolean).

andrewbenton committed:
Use override method to set if bsync is enabled.

The getter/setter with locking added more to the API than was necessary.
 This alternative method only adds a single method that must be
 overridden in order for the programmer to enable barrier
 synchronization for the test class.
